### PR TITLE
Negative case for malformed mhc-yaml(from v4.5) OCP-29857

### PR DESCRIPTION
**Pull request by miyadav**
Negative case for malformed mhc-yaml(from v4.5) , Maxunhealthy should not allow invalid value - OCP29857

Run output -
      [12:53:38] INFO> Exit Status: 0
      [12:53:38] INFO> Shell Commands: oc delete machines machineset-clone-29857-25mhf --wait=false --config=/root/workdir/07e4c26efebb-/ocp4_admin.kubeconfig --namespace=openshift-machine-api
      machine.machine.openshift.io "machineset-clone-29857-25mhf" deleted
      
      [12:53:38] INFO> Exit Status: 0
      [12:53:39] INFO> oc get machines machineset-clone-29857-25mhf --output=yaml --config=/root/workdir/07e4c26efebb-/ocp4_admin.kubeconfig --namespace=openshift-machine-api
      [13:01:18] INFO> After 163 iterations and 459 seconds:
      
      STDERR:
      Error from server (NotFound): machines.machine.openshift.io "machineset-clone-29857-25mhf" not found
      
      [13:01:18] INFO> Shell Commands: oc get machines machineset-clone-29857-25mhf --output=yaml --config=/root/workdir/07e4c26efebb-/ocp4_admin.kubeconfig --namespace=openshift-machine-api
      
      STDERR:
      Error from server (NotFound): machines.machine.openshift.io "machineset-clone-29857-25mhf" not found
      
      [13:01:21] INFO> Exit Status: 1
      [13:01:21] INFO> Shell Commands: rm -r -f -- /root/workdir/07e4c26efebb-
      
      [13:01:21] INFO> Exit Status: 0
      [13:01:21] INFO> === End After Scenario: [MHC] MaxUnhealthy should not allow malformed values ===

1 scenario (1 passed)
15 steps (15 passed)
16m17.257s
[13:01:21] INFO> === At Exit ===

### CODE REVIEW

**Comment by miyadav** [comment]
@jhou1  @sunzhaohua2 PTAL tomm

File: features/machine/machine_healthcheck.feature [hunk 1]
```diff
@@ -166,5 +166,44 @@ Feature: MachineHealthCheck Test Scenarios
      | Ensuring a requeue happens in 15m0  |
     """
    Then the machine should be remediated
+ 
+  # @author miyadav@redhat.com
+  # @case_id OCP-29857
+  @admin
+  @destructive
```
Comment (jhou1) [inline features/machine/machine_healthcheck.feature:173]: Since we are creating a malformed mhc, we don't really have to clone a machineset. We can create the mhc against a current machineset and check the mhc log directly.  the `@destructive` tag can be removed

File: features/machine/machine_healthcheck.feature [hunk 2]
```diff
@@ -166,5 +166,44 @@ Feature: MachineHealthCheck Test Scenarios
...
+  Scenario: [MHC] MaxUnhealthy should not allow malformed values
+    Given I have an IPI deployment
+    And I switch to cluster admin pseudo user
+    Then I use the "openshift-machine-api" project
+    And admin ensures machine number is restored after scenario
+
+    Given I clone a machineset and name it "machineset-clone-29857"
```
Comment (jhou1) [inline features/machine/machine_healthcheck.feature:180]: No need to clone a new machineset here

File: features/machine/machine_healthcheck.feature [hunk 3]
```diff
@@ -166,5 +166,44 @@ Feature: MachineHealthCheck Test Scenarios
...
+
+    # Create MHC with malformed unhealthy nodes value
+    When I run oc create over "<%= BushSlicer::HOME %>/testdata/cloud/mhc/mhc_malformed.yaml" replacing paths:
```
Comment (jhou1) [inline features/machine/machine_healthcheck.feature:183]: Replace this line with
```
Given I obtain test data file "xxx"
When I run oc create over "xxx" replacing paths:
```

File: features/machine/machine_healthcheck.feature [hunk 4]
```diff
@@ -166,5 +166,44 @@ Feature: MachineHealthCheck Test Scenarios
...
+     | ["spec"]["selector"]["matchLabels"]["machine.openshift.io/cluster-api-cluster"]    | <%= machine_set.cluster %>  |
+     | ["spec"]["selector"]["matchLabels"]["machine.openshift.io/cluster-api-machineset"] | <%= machine_set.name %>     |
+    Then the step should succeed
+    And I ensure "mhc-<%= machine_set.name %>" machinehealthcheck is deleted after scenario
+ 
+    Given I create the 'Ready' unhealthyCondition
+    Then a pod becomes ready with labels:
+     | api=clusterapi, k8s-app=controller |
+
+     When I run the :logs admin command with:
```
Comment (jhou1) [inline features/machine/machine_healthcheck.feature:195]: Fix the indent

File: features/machine/machine_healthcheck.feature [hunk 5]
```diff
@@ -166,5 +166,44 @@ Feature: MachineHealthCheck Test Scenarios
...
+    # Create MHC with malformed unhealthy nodes value
+    When I run oc create over "<%= BushSlicer::HOME %>/testdata/cloud/mhc/mhc_malformed.yaml" replacing paths:
+     | n                                                                                  | openshift-machine-api       |
+     | ["metadata"]["name"]                                                               | mhc-<%= machine_set.name %> |
+     | ["spec"]["selector"]["matchLabels"]["machine.openshift.io/cluster-api-cluster"]    | <%= machine_set.cluster %>  |
+     | ["spec"]["selector"]["matchLabels"]["machine.openshift.io/cluster-api-machineset"] | <%= machine_set.name %>     |
+    Then the step should succeed
+    And I ensure "mhc-<%= machine_set.name %>" machinehealthcheck is deleted after scenario
+ 
+    Given I create the 'Ready' unhealthyCondition
```
Comment (jhou1) [inline features/machine/machine_healthcheck.feature:191]: No need to create the unhealthycondition as well. You can just run the next step to see the log

File: features/machine/machine_healthcheck.feature [hunk 6]
```diff
@@ -166,5 +166,44 @@ Feature: MachineHealthCheck Test Scenarios
...
+    Then a pod becomes ready with labels:
+     | api=clusterapi, k8s-app=controller |
+
+     When I run the :logs admin command with:
+     | resource_name | <%= pod.name %>                |
+     | c             | machine-healthcheck-controller |
+    Then the output should contain:
+     | invalid value for IntOrString |
```
Comment (jhou1) [inline features/machine/machine_healthcheck.feature:199]: Replace with `| invalid value "3%0" |` looks better

File: testdata/cloud/mhc/mhc_malformed.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+apiVersion: "machine.openshift.io/v1beta1"
+kind: "MachineHealthCheck"
+metadata:
+  name: mhc1
```
Comment (jhou1) [inline testdata/cloud/mhc/mhc_malformed.yaml:4]: suggest to rename to *mhc_mailformed* or something similar

File: features/machine/machine_healthcheck.feature [hunk 8]
```diff
@@ -166,5 +166,44 @@ Feature: MachineHealthCheck Test Scenarios
...
+ 
+  # @author miyadav@redhat.com
+  # @case_id OCP-29857
+  @admin
+  @destructive
+  Scenario: [MHC] MaxUnhealthy should not allow malformed values
+    Given I have an IPI deployment
+    And I switch to cluster admin pseudo user
+    Then I use the "openshift-machine-api" project
+    And admin ensures machine number is restored after scenario
```
Comment (jhou1) [inline features/machine/machine_healthcheck.feature:178]: Remove this step as we are not going to scale the nodes

**Comment by miyadav** [comment]
Thanks @jhou1  , for review comments , I made better modification to include other test case as well , here is the validation output , Included empty selectors in mhc test case as well in this one.
PTAL
.
.
  [04:28:30] INFO> Exit Status: 0
      | resource_name | <%= pod.name %>                |
      | c             | machine-healthcheck-controller |
    Then the output should contain:                                                                                                           # features/step_definitions/common.rb:33
      | remediation won't be allowed: invalid value for IntOrString |
      | total targets: <%= cb.machines.count %>                     |
waiting for operation up to 3600 seconds..
waiting for operation up to 300 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
      [04:28:30] INFO> === After Scenario: [MHC] MaxUnhealthy should not allow malformed values ===
      [04:28:30] INFO> Shell Commands: oc delete machinehealthcheck mhc-malformed --wait=false --config=/root/workdir/07e4c26efebb-/ocp4_admin.kubeconfig --namespace=openshift-machine-api
      machinehealthcheck.machine.openshift.io "mhc-malformed" deleted
      
      [04:28:31] INFO> Exit Status: 0
      [04:28:32] INFO> oc get machinehealthcheck mhc-malformed --output=yaml --config=/root/workdir/07e4c26efebb-/ocp4_admin.kubeconfig --namespace=openshift-machine-api
      [04:28:32] INFO> After 1 iterations and 0 seconds:
      
      STDERR:
      Error from server (NotFound): machinehealthchecks.machine.openshift.io "mhc-malformed" not found
      
      [04:28:32] INFO> Shell Commands: rm -r -f -- /root/workdir/07e4c26efebb-
      
      [04:28:32] INFO> Exit Status: 0
      [04:28:32] INFO> === End After Scenario: [MHC] MaxUnhealthy should not allow malformed values ===

1 scenario (1 passed)
10 steps (10 passed)
0m15.438s
[04:28:32] INFO> === At Exit ===

File: features/machine/machine_healthcheck.feature (hunk 3 above)
Comment (miyadav) [reply features/machine/machine_healthcheck.feature:180]: since I am only changing namespace ,do i need to obtain file or this way is also ok ?

File: features/machine/machine_healthcheck.feature (hunk 6 above)
Comment (miyadav) [reply features/machine/machine_healthcheck.feature]: this comes in two lines hence not using it

File: features/machine/machine_healthcheck.feature (hunk 3 above)
Comment (jhou1) [reply features/machine/machine_healthcheck.feature:180]: I was suggested to use the above steps in a PR several days back, so I thought that had some advantage, I didn't look into their step definitions though. @liangxia Can you please suggest?

**Comment by jhou1** [comment]
Thank you @miyadav , looks good to me. Please also merge the two test cases in polarion.

File: features/machine/machine_healthcheck.feature (hunk 3 above)
Comment (pruan-rht) [reply features/machine/machine_healthcheck.feature:180]: @jhou1 @miyadav the even though it's a two steps process, there are two advantages

1. getting data files in abstracted way so that if in the future `BushSlicer::HOME` changes we only need to change in the `I obtain test data file` step.
2. less typing in the actual

Comment (pruan-rht) [reply features/machine/machine_healthcheck.feature:180]: @miyadav plese use the `I obtain test data file...` approach...thx!

Comment (miyadav) [reply features/machine/machine_healthcheck.feature]: got it added this

**Comment by jhou1** [comment]
There are conflicts, need rebase.

**Comment by miyadav** [comment]
run after new commit- 
      [07:09:45] INFO> Exit Status: 0
      [07:09:45] INFO> === End After Scenario: [MHC] MaxUnhealthy should not allow malformed values ===

1 scenario (1 passed)
11 steps (11 passed)
0m19.992s
[07:09:45] INFO> === At Exit ===
